src/metabase/pulse.clj

```diff
@@ -261,24 +264,35 @@
          :email (:email recipient)}
         {:kind :user
          :user recipient}))
+    :http
+    []
     (do
      (log/warnf "Unknown channel type %s" (:channel_type pulse-channel))
      [])))
 
-(defn- channel-send!
-  [& args]
-  (try
-    (apply channel/send! args)
-    (catch Exception e
-      ;; Token errors have already been logged and we should not retry.
-      (when-not (and (= :channel/slack (:type (first args)))
-                     (contains? (:errors (ex-data e)) :slack-token))
-        (throw e)))))
-
 (defn- send-retrying!
-  [& args]
+  [pulse-id channel message]
   (try
-    (apply (retry/decorate channel-send!) args)
+    (let [;; once we upgraded to retry 2.x, we can use (.. retry getMetrics getNumberOfTotalCalls) instead of tracking
+          ;; this manually
+          retry-count (volatile! 0)
+          send!       (fn []
+                          (task-history/with-task-history {:task         "channel-send"
+                                                           :task_details {:retry-config   (when retry/*retry-config*
```

Comment:
this is subjective but I'm not a huge fan of using dynamic scope if it can be easily avoided because it's much harder to follow. this falls in the "easily avoided" bucket imo.

something like this would work:

```clojure
(defn- send-retrying!
  [pulse-id channel message]
  (try
    (let [;; once we upgraded to retry 2.x, we can use (.. retry getMetrics getNumberOfTotalCalls) instead of tracking
          ;; this manually
          retry-count  (volatile! 0)
          retry-config (retry/retry-configuration)
          send!        (fn []
                         (task-history/with-task-history {:task         "channel-send"
                                                          :task_details {:max-attempts   (:max-attempts retry-config)
                                                                         :channel-type   (:type channel)
                                                                         :channel-id     (:id channel)
                                                                         :pulse-id       pulse-id
                                                                         :retry-attempts @retry-count}}
                           (try
                             (channel/send! channel message)
                             (catch Exception e
                               (vswap! retry-count inc)
                                ;; Token errors have already been logged and we should not retry.
                               (when-not (and (= :channel/slack (:type channel))
                                              (contains? (:errors (ex-data e)) :slack-token))
                                 (throw e))))))]
      ((retry/decorate send! (retry/random-exponential-backoff-retry (str (random-uuid)) retry-config))))
    (catch Throwable e
      (log/error e "Error sending notification!"))))
```

Reply:
I also replaced `:retry-config` with just `:max-attempts` above

Reply:
+1

Reply:
Well, +1 for sentiment, but it looks like the config can be rebound in other places so just calling `(retry/retry-configuration)` won't work. Might have to add it to the parameters

Reply:
updated